src/adt/edgeset.c

```diff
@@ -871,3 +887,997 @@ edge_set_ordered_iter_next(struct edge_ordered_iter *eoi, struct fsm_edge *e)
 		}
 	}
 }
+
+#else /* USE_EDGE_BITSET */
+
+#define LOG_BITSET 0
+
+#include "libfsm/internal.h" /* XXX: for allocating struct fsm_edge, and the edges array */
+
+#include <print/esc.h>
+
+#include <adt/alloc.h>
+#include <adt/bitmap.h>
+#include <adt/set.h>
+#include <adt/stateset.h>
+#include <adt/edgeset.h>
+
+#define DEF_EDGE_GROUP_CEIL 2
+
+/* Array of <to, symbols> tuples, sorted by to.
+ *
+ * Design assumption: It is significantly more likely in practice to
+ * have to be more edges with different labels going to the same state
+ * than the same symbol going to different states. This does not
+ * include epsilon edges, which can be stored in a state_set.
+ *
+ * A NULL pointer is treated as the empty set, because */
+struct edge_set {
+	size_t ceil;		/* nonzero */
+	size_t count;		/* <= ceil */
+	struct edge_group {
+		fsm_state_t to;	/* distinct */
+		uint64_t symbols[256/64];
+	} *groups;		/* sorted by .to */
+};
+
+#define SYMBOLS_SET(S, ID) (S[ID/64] |= (1ULL << (ID & 63)))
+#define SYMBOLS_GET(S, ID) (S[ID/64] & (1ULL << (ID & 63)))
+#define SYMBOLS_CLEAR(S, ID) (S[ID/64] &=~ (1ULL << (ID & 63)))
+
+struct edge_set *
+edge_set_new(void)
+{
+#if LOG_BITSET
+	fprintf(stderr, " -- edge_set_new\n");
+#endif
+	return NULL;		/* -> empty set */
+}
+
+void
+edge_set_free(const struct fsm_alloc *a, struct edge_set *set)
+{
+#if LOG_BITSET
+	fprintf(stderr, " -- edge_set_free %p\n", (void *)set);
+#endif
+	if (set == NULL) {
+		return;
+	}
+
+	f_free(a, set->groups);
+	f_free(a, set);
+}
+
+static int
+grow_groups(struct edge_set *set, const struct fsm_alloc *alloc)
+{
+	/* TODO: This could also squash out any groups where
+	 * none of the symbols are set anymore. */
+	const size_t nceil = 2 *set->ceil;
+	struct edge_group *ng;
+	assert(nceil > 0);
+
+#if LOG_BITSET
+	fprintf(stderr, " -- edge_set grow_groups: %lu -> %lu\n",
+	    set->ceil, nceil);
+#endif
+
+	ng = f_realloc(alloc, set->groups,
+	    nceil * sizeof(set->groups[0]));
+	if (ng == NULL) {
+		return 0;
+	}
+
+	set->ceil = nceil;
+	set->groups = ng;
+
+	return 1;
+}
+
+static void
+dump_edge_set(const struct edge_set *set)
+{
+	const struct edge_group *eg;
+	size_t i;
+#if LOG_BITSET < 2
+	return;
```

Comment:
Do you have a preferred naming convention for this kind of thing? It logs the struct, but only if the log level is sufficiently high. `maybe_dump_edge_set`? `dump_edge_set_if_logging`?

Reply:
No preference from me. "dump" sounds fine (I tend to use that for datastructures).